Feature | Add scaffolding for SSRP response parsing

edwardneal opened the pull request:
## Description

This is the first of four or five PRs which introduce an SSRP client to the managed SNI. This PR doesn't actually introduce any parsing code - it just adds test cases and lays some scaffolding/infrastructure down for the next one.

Key points of interest here:
* Around two thirds of the additions here are test cases. I've tried to cover as many of these as I can see based upon the MC-SQLR specification, with a particular focus on malformed data. These will end up parsing unsanitised network traffic, so I'm completely open to adding any others which would cover malicious traffic.
* I'm planning to make fairly heavy use of `ReadOnlySequence<byte>` to store packet buffers, then read the SSRP responses from them. This introduces the use of `PacketBuffer` and `ReadOnlySequenceUtilities`.
* The existence of `ReadOnlySequenceUtilities` is a workaround; netcore has a [`SequenceReader<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.buffers.sequencereader-1) type in-box. This isn't available to netfx though, and I didn't think it was worth maintaining two sets of fairly simple logic. We could alternatively bring the `SequenceReader<T>` and `SequenceReaderExtensions` source code in-tree, as a few other libraries have done (examples [1](https://github.com/MessagePack-CSharp/MessagePack-CSharp/blob/master/src/MessagePack/SequenceReader.cs) and [2](https://github.com/dotnet/Nerdbank.Streams/blob/main/src/Nerdbank.Streams/SequenceReader.cs).)

## Issues

Contributes to #3700.

## Testing

There's nothing to test in this PR, so CI should validate it.

edwardneal committed:
Add scaffolding for SSRP response parsing

This consists of:
* ReadOnlySequenceSegment derivative,
* Utilities to read data from a ReadOnlySequence<byte>
* Test cases

apoorvdeshmukh commented:
/azp run

edwardneal commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/Common/ReadOnlySequenceUtilities.cs:
```diff
@@ -0,0 +1,92 @@
+    /// <returns><c>true</c> if <paramref name="sequence"/> was long enough to retrieve the next two bytes, <c>false</c> otherwise.</returns>
+    public static bool ReadLittleEndian(this ref ReadOnlySequence<byte> sequence, ref ReadOnlySpan<byte> currSpan, ref long currPos, out ushort value)
+    {
+        if (sequence.Length < sizeof(ushort))
```
As per the ReadByte method - `currPos` is always in a consistent state. Once execution reaches line 69, there's guaranteed to be enough space in the `ReadOnlySpan<byte>` - the only question is whether we can directly read it from the current span, or need to reassemble it because byte 1 is in the current span and byte 2 is in the next span.

edwardneal commented:
This PR is not stale.

paulmedynski commented:
Since this is a new feature, we will need to prioritize our review of it against existing priorities, so this may not see much attention until early 2026.  @edwardneal Do you think you could have all of the work for SSRP ready for the 7.0 GA in March?

edwardneal commented:
Thanks @paulmedynski. I tried rebasing, but that went badly wrong and the CodeQL error persisted. I think I've fixed it now, and the build looks like it's triggered properly.

March is enough time to implement the feature, and I have a local branch with the SSRP response parsing logic and test cases. I definitely wouldn't prioritise SqlDataSourceEnumerator above the other changes in 7.0 though, and I think it's likely to be a little more sensitive to review (since the only way it can work is by parsing untrusted network broadcasts.)

I'm happy enough for this series of PRs to lie dormant until post-GA. Implementation doesn't introduce a breaking change, it just means that clients can enumerate SQL Server instances on the local LAN rather than having GetDataSources throw an exception.

edwardneal commented:
This PR is not stale.

paulmedynski commented:
@edwardneal - New conflicts here.  I will work to get this approved, reviewed, and merged ASAP to avoid wasting your time with further conflicts.

edwardneal committed:
Merge main

edwardneal commented:
Thanks @paulmedynski, I've just merged.

Most of the conflicts have come from PRs which predate the project merge. I'm not expecting many conflicts in this area of the codebase moving forwards, so we don't need to rush and displace any other priorities on that basis.

benrr101 commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/Common/ReadOnlySequenceUtilities.cs:
```diff
@@ -0,0 +1,92 @@
+    /// <param name="currPos">Current position in the sequence. Advanced by two bytes following a successful read.</param>
+    /// <param name="value">The <see cref="ushort"/> value read from <paramref name="sequence"/></param>
+    /// <returns><c>true</c> if <paramref name="sequence"/> was long enough to retrieve the next two bytes, <c>false</c> otherwise.</returns>
+    public static bool ReadLittleEndian(this ref ReadOnlySequence<byte> sequence, ref ReadOnlySpan<byte> currSpan, ref long currPos, out ushort value)
```
Unless this is attempting to match an existing API, I'd prefer it if the type we're reading is in the name. Just so that if we add more methods for reading like this, we know what type we're reading.

So, this one would be `ReadUshortLittleEndian`

benrr101 replied:
Or `ReadUInt16LittleEndian` - your choice

edwardneal replied:
While I agree and would normally like to align this with BinaryPrimitives (i.e. ReadUInt16LittleEndian), this is trying to align with an existing API - [SequenceReaderExtensions.TryReadLittleEndian](https://learn.microsoft.com/en-gb/dotnet/api/system.buffers.sequencereaderextensions.tryreadlittleendian).

The method signature is different, so I'm not particularly fixated on keeping the same name though.

benrr101 commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/Common/ReadOnlySequenceUtilities.cs:
```diff
@@ -0,0 +1,92 @@
+            Span<byte> buffer = stackalloc byte[sizeof(ushort)];
+
+            sequence.Slice(0, sizeof(ushort)).CopyTo(buffer);
+            value = BinaryPrimitives.ReadUInt16LittleEndian(buffer);
```
I'd double check the behavior of ReadUInt16LittleEndian to see if it already does this partial read behavior. (Or I could just be mixing this up with some of the stuff I wrote for a different project ...)

edwardneal replied:
It doesn't, unfortunately. As per [documentation](https://learn.microsoft.com/en-us/dotnet/api/system.buffers.binary.binaryprimitives.readuint16littleendian), it throws an exception if the source buffer is too small to contain a `ushort`.

benrr101 commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/Sql/DacResponseProcessorTest.cs:
```diff
@@ -0,0 +1,36 @@
+    [ActiveIssue("https://github.com/dotnet/SqlClient/issues/3700")]
+    public void Process_EmptyBuffer_ReturnsFalse(ReadOnlySequence<byte> packetBuffers)
+    {
+        _ = packetBuffers;
```
I assume these are place holders for something else? If so, can we get a comment here indicating such?
The active issue attribute is being used by us to indicate when a test is broken due to an open issue.

edwardneal replied:
These are placeholders for the test cases. I've swapped ActiveIssue with the Skip property on Theory.

benrr101 commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/Sql/SsrpPacketTestData.cs:
```diff
@@ -0,0 +1,501 @@
+    {
+        get
+        {
+            byte[] complexValidPacket = FormatSVR_RESPMessage(0x05,
```
Please make sure if you're chopping a line all elements should go on a new line.

```
byte[] complexValidPacket = FormatSvr_RespMessage(
  0x05,
  CreateResp_Data(
    "svr1",
    "MSSQLSERVER",
    true,
    "14.0.0.0",
    CreateProtocolParameters(
      tcpInfo: "tcp;1433",
      npInfo: @"np;\\svr1\pipe\SamplewithanaccentPipeName",
      viaInfo: "via;svr1 1:1433",
      rpcInfo: "rpc;svr1",
      spxInfo: "spx;MSSQLSERVER",
      adspInfo: "adsp;SQL2000",
      bvInfo: "bv;item;group;item;group;org",
      otherParameters: null)));
```

if that's too ugly ... feel free to refactor to make it cleaner :)

edwardneal replied:
I've performed lots of line chopping and made slightly better use of optional parameters.

benrr101 commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/Sql/SsrpPacketTestData.cs:
```diff
@@ -0,0 +1,501 @@
+            byte[] validPacket1 = FormatSVR_RESPMessage(0x05,
+                CreateRESP_DATA("svr1", "MSSQLSERVER", true, "14.0.0.0",
+                    CreateProtocolParameters(tcpInfo: "tcp;1433", null, null, null, null, null, null, null)));
+            byte[] validPacket2 = FormatSVR_RESPMessage(0x05,
```
It would be preferable to have all constants that aren't obvious what they do to have a parameter label. Especially nulls!

edwardneal replied:
Yes - lifted to constants, and made more liberal use of parameter labels.

benrr101 commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/Sql/SsrpPacketTestData.cs:
```diff
@@ -0,0 +1,501 @@
+            return new()
+            {
+                // One buffer, one response
+                { GeneratePacketBuffers(
```
Please inline things where they can be inlined cleanly. And if they can't be and must be chopped, put all elements on new lines.

```
{ GeneratePacketBuffers(complexValidPacket), "14.0.0.0", 1433, @"\\svr1\pipe\SamplewithanaccentPipeName" },
{
    GeneratePacketBuffers(
      complexValidPacket.AsSpan(0, 14).ToArray(),
      ...),
    "14.0.0.0",
    1433,
    @"\\svr1\pipe\SamplewithanaccentPipeName"
},
...
```

benrr101 commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/Sql/SsrpPacketTestData.cs:
```diff
@@ -0,0 +1,501 @@
+    /// should be successfully processed.
+    /// </summary>
+    /// <see cref="SqlDataSourceResponseProcessorTest.Process_ValidDacResponse_ReturnsData"/>
+    public static TheoryData<ReadOnlySequence<byte>, int> ValidSVR_RESP_DACPacketBuffer
```
I get that packet is a SVR_RESP_DAC response, but I'm gonna be a stickler and say we need to stick to C# naming conventions. So, this should be `ValidSvrRespDacResponse` or something less abbreviated if that'd read better.

edwardneal replied:
Yes, done.

benrr101 commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/Sql/SsrpPacketTestData.cs:
```diff
@@ -0,0 +1,501 @@
+    /// <see cref="DacResponseProcessorTest.Process_InvalidDacResponse_ReturnsFalse"/>
+    public static TheoryData<ReadOnlySequence<byte>> InvalidSVR_RESP_DACPackets
+    {
+        get
```
Feel free, if it makes sense and isn't too ugly to use expression bodies for these

edwardneal replied:
I can only practically use it here, but it's in place now.

benrr101 commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/Sql/SsrpPacketTestData.cs:
```diff
@@ -0,0 +1,501 @@
+                GeneratePacketBuffers(
+                    FormatSVR_RESPMessage(0x05,
+                        CreateRESP_DATA("srv1", "MSSQLSERVER", true, "14.0.0.0",
+                            CreateProtocolParameters(tcpInfo, null, null, null, null, null, null, null)
```
Don't forget the standard is to put all closing ) on the same line
(Don't ask me, I didn't make the standard ... and I generally don't prefer it either)

edwardneal replied:
Done!

benrr101 commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/Sql/SsrpPacketTestData.cs:
```diff
@@ -0,0 +1,501 @@
+    /// <param name="bvInfo">If non-null, the BV_INFO data.</param>
+    /// <param name="otherParameters">Any additional protocol parameters to include.</param>
+    /// <returns>The collated protocol parameters for a RESP_DATA section.</returns>
+    private static string CreateProtocolParameters(string? tcpInfo, string? npInfo,
```
Please chop all arguments if they can't fit on one line

edwardneal replied:
Arguments chopped (and given default values of null, where appropriate.)

benrr101 reviewed (changes requested):
I'm not entirely sure how this all comes together, buuuuuut I think it's fine. Requesting a handful of style changes for the tests, and a couple things to double check before approving.

edwardneal committed:
Merge branch 'main' into feat/ssrp/01-scaffolding

edwardneal committed:
Replace ActiveIssue attributes with [Theory(Skip = "...")]

edwardneal committed:
Formatting changes

Lifted some constant string components to constants.
Newline formatting.
Where possible, used expression body.

edwardneal committed:
Naming convention changes

edwardneal committed:
Parameter line chopping

edwardneal committed:
Respond to code review

Correct screaming snake case method.
Correct ensure that test case provides a zero-length Version field (not an absent one.)

edwardneal committed:
Add SSRP parsing test cases

edwardneal commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/Sql/SsrpPacketTestData.cs:
```diff
@@ -0,0 +1,644 @@
+                        isClustered: true,
+                        version: string.Empty,
+                        validTcpInfo,
+                        omitVersion: true))),
```
Yes - fixed.

edwardneal commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/Sql/SsrpPacketTestData.cs:
```diff
@@ -0,0 +1,644 @@
+                CreateSVR_RESPMessage("tcp;-1")
+            ];
+
+            static ReadOnlySequence<byte> CreateSVR_RESPMessage(string tcpInfo) =>
```
SCREAMING_SNAKE_CASE has now been removed.

benrr101 reviewed (approved):
Thanks for the changes! Looks good to me!

benrr101 committed:
Merge branch 'main' into feat/ssrp/01-scaffolding
